### PR TITLE
Attach finalizers to LibGit2 methods

### LINKED ISSUE #17414: LibGit2 should not extend `Base.finalize`
`Base.finalize` has well define meaning and the overload doesn't do that.

Ref https://github.com/JuliaLang/julia/pull/17409#discussion_r70791075

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #17414.

### CODE REVIEW

**Comment by simonbyrne** [comment]
The alternative here, of course, is to actually add these methods as finalizers to the relevant objects: this would probably make more sense in the long run (as a benefit, we wouldn't need to then deprecate the `finalize` methods, as these would be implicit).

**Review by tkelman — APPROVED** [review]
this lgtm for now. be sure to squash

**Comment by simonbyrne** [comment]
@wildart Do you know why we don't use object finalizers for these?

**Comment by simonbyrne** [comment]
I've cleaned this up so that most now actually act like finalizers. This can't be used for `StrArrayStruct` and `Buffer` though, as they're immutable, but these aren't really user facing.

**Comment by tkelman** [comment]
seems to cause Pkg issues, see the travis log

**Comment by simonbyrne** [comment]
Okay, hopefully this should be good.

File: base/libgit2/types.jl [hunk 1]
```diff
@@ -373,53 +373,100 @@ abstract AbstractGitObject
 Base.isempty(obj::AbstractGitObject) = (obj.ptr == C_NULL)
 
 abstract GitObject <: AbstractGitObject
-function Base.finalize(obj::GitObject)
-    if obj.ptr != C_NULL
-        ccall((:git_object_free, :libgit2), Void, (Ptr{Void},), obj.ptr)
-        obj.ptr = C_NULL
+
+macro defclose(typ,cname)
```
Comment (tkelman) [inline base/libgit2/types.jl]: is this used anywhere?

File: base/libgit2/types.jl [hunk 2]
```diff
@@ -373,54 +373,100 @@ abstract AbstractGitObject
...
+    (:GitRebase,     :GitRepo,  :AbstractGitObject, :git_rebase),
+    (:GitStatus,     :GitRepo,  :AbstractGitObject, :git_status_list),
+    (:GitBranchIter, :GitRepo,  :AbstractGitObject, :git_branch_iterator),
+    (:GitAnyObject,  :GitRepo,  :GitObject,         :git_object),
+    (:GitCommit,     :GitRepo,  :GitObject,         :git_commit),
+    (:GitBlob,       :GitRepo,  :GitObject,         :git_blob),
+    (:GitTree,       :GitRepo,  :GitObject,         :git_tree),
+    (:GitTag,        :GitRepo,  :GitObject,         :git_tag)]
+
+    if reporef == nothing
```
Comment (tkelman) [inline base/libgit2/types.jl]: `===`

File: base/libgit2/types.jl (hunk 1 above)
Comment (simonbyrne) [inline base/libgit2/types.jl]: ah, no: was left over from my refactoring

File: base/libgit2/types.jl (hunk 2 above)
Comment (simonbyrne) [inline base/libgit2/types.jl]: fixed.

File: base/libgit2/libgit2.jl [hunk 3]
```diff
@@ -432,25 +432,25 @@ function merge!(repo::GitRepo;
...
             else
-                with(head(repo)) do head_ref
-                    with(upstream(head_ref)) do tr_brn_ref
-                        if tr_brn_ref === nothing
-                            throw(GitError(Error.Merge, Error.ERROR,
-                                           "There is no tracking information for the current branch."))
-                        end
-                        [GitAnnotated(repo, tr_brn_ref)]
-                    end
+                tr_brn_ref = upstream(head_ref)
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: `upstream(head(repo))` ?

Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: Hmm, something go broken in the rebase. I'll try to figure out what happened.

**Comment by simonbyrne** [comment]
Okay, I think this is ready.

**Comment by KristofferC** [comment]
Is there any requirement from libgit2 that objects should be freed "quickly" as they are when using the `with` form? Or is it fine to just delay the freeing until whenever the finalizers are run?

**Comment by simonbyrne** [comment]
Not that I can tell, no.

In fact, we could probably clean up the code quite a bit by getting rid of all the `with` statements.

**Comment by tkelman** [comment]
Unless I guess anything on the C side holds open file handles that we'd like to delete or similar?

**Comment by simonbyrne** [comment]
That could be, but the libgit2 docs are pretty sparse in that regard.

**Pull request merged**